qcelemental/models/objects/molecule.py

```diff
@@ -0,0 +1,107 @@
+"""
+Molecule Object Model
+"""
+
+import numpy as np
+from pydantic import BaseModel, validator
+from typing import List, Tuple
+
+
+class NPArray(np.ndarray):
+    @classmethod
+    def get_validators(cls):
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, v):
+        try:
+            v = np.array(v)
+        except:
+            raise RuntimeError("Could not cast {} to NumPy Array!".format(v))
+        return v
+
+class Identifiers(BaseModel):
+    """Canonical chemical identifiers"""
+    molecule_hash: str = None
+    molecular_formula: str = None
+    smiles: str = None
+    inchi: str = None
+    inchikey: str = None
+    canonical_explicit_hydrogen_smiles: str = None
+    canonical_isomeric_explicit_hydrogen_mapped_smiles: str = None
+    canonical_isomeric_explicit_hydrogen_smiles: str = None
+    canonical_isomeric_smiles: str = None
+    canonical_smiles: str = None
+
+    class Config:
+        allow_extra = True
+
+
+# Outstanding issues:
+# required_definitions
+# hash_fields
+# Numpy arrays or lists?
+class Molecule(BaseModel):
+    id: str = None
+    symbols: List[str]
+    geometry: NPArray = ...
+    masses: List[float] = None
+    name: str = None
+    identifiers: Identifiers = None
+    comment: str = None
+    charge: float = 0.0
+    multiplicity: int = 1
+    real: List[bool] = None
+    connectivity: List[Tuple[int, int, int]] = []
+    fragments: List[List[int]] = None
+    fragment_charges: List[float] = None
+    fragment_multiplicities: List[int] = None
+    fix_com: bool = False
+    fix_orientation: bool = False
+    provenance: dict = {}
+
+    @validator('geometry')
+    def must_be_3n(cls, v, values, **kwargs):
```

Comment:
Before we go too deep on this kind of validation with pydantic itself, @loriab do we have validation routines that can be used here?

Reply:
for testing, there's https://github.com/MolSSI/QCElemental/blob/master/qcelemental/tests/utils_compare.py#L134 that after `from utils import *` is ready to be used with `assert`.

Reply:
I'd very much recommend stuffing all these parameters into [molparse.from_arrays](https://github.com/MolSSI/QCElemental/blob/master/qcelemental/molparse/from_arrays.py#L130). That'll keep all the validation code localized and also validate self-consistently, rather than field-by-field.

Reply:
Recommend staying away from _any_ `from utils import *`

Reply:
agree in general. this is testing only and a [tight little file](https://github.com/MolSSI/QCElemental/blob/master/qcelemental/tests/utils.py), so i've considered it harmless. now that I look at it, I should privatize the `true_false_decorator`.